enhancements/network/egress-ip-per-destinations.md

```diff
@@ -0,0 +1,327 @@
+see-also:
+- "NA"
+replaces:
+- "NA"
+superseded-by:
+- "NA"
+---
+
+# Egress IP per destination
+## Summary
+Today, we can use an `EgressIP` to describe the source IP for a pod if it selected by an `EgressIP` custom resource (CR) selectors. This includes namespace
+and pod selectors. If multiple `EgressIP` CRs select the same set of pods, the behavior is undefined. This is because
+we cannot reliably choose which source IP to use.
+
+This enhancement proposes adding a new selector for when we want to apply the `EgressIP`. This new selector will only apply
+the `EgressIP` as the source IP to a set of pods communicating with destination IP if that destination IP is within predefined network CIDRs.
+
+If the new destination traffic selector is specified for all `EgressIP` CRs that selected a set of pods and the destination networks selected do not overlap, then we
+can allow a set of pods to have multiple source IPs depending on the destination address.
+
+## Motivation
+### User Stories
+
+As a cluster tenant, I want my application traffic to egress a particular interface depending on the traffic destination address and must
+have a distinct but consistent source IP for each interface.
+
+As a cluster administrator, I want to easily configure which destination networks should have an `EgressIP` because the destination
+networks may change, and I do not wish to have to update individual `EgressIP` custom resources for every update.
+
+### Goals
+- Allow a set of pods to have a consistent source IP depending on the destination
+- Support shared and local gateway (aka `routingViaHost=true`) modes in-order to reduce asymmetry between the modes
+- Backward compatibility with existing `EgressIP` API
+- If the new API isn't defined, preserve existing behaviour
+
+### Non-Goals
+- Change existing `EgressIP` behavior when the new field is not defined
+
+## Proposal
+
+A new `TrafficSelector` field is added to `EgressIP` spec which uses Kubernetes label selector semantics
+to select a new CRD called `EgressIPTraffic` which contains a list of destination networks. This list of networks defines when an `EgressIP` should
+be used as source IP when an application attempts to talk to a destination address, and it is within one of the predefined networks
+defined within a `EgressIPTraffic` CR.
+
+The reason for not embedding the destination networks within each `EgressIP` CR is that it places a burden
+on the cluster administrator to update all `EgressIP` CRs if the destination networks change.
+
+With the Kubernetes label selector semantics, cluster administrators can define a set of destination networks that multiple `EgressIP`
+CRs can consume.
+
+If greater than one `EgressIPTraffic` CRs are selected, the networks are added together, and if a destination IP is within one
+of the networks, then the pods source IP will be the `EgressIP`.
+If `TrafficSelector` is not specific, existing `EgressIP` behaviour is conserved.
+
```

Comment:
That scenario would be undefined - either `EgressIP` CR can be "active" but not both because EgressIP B would select all traffic. 

Today, if 2 `EgressIP`s select the same set of pods, one is active and the other is a standby.

With this enhancement, we allow overlapping (i.e. select set of one or more pods) `EgressIP` CRs only if the selected `EgressIPTraffic` destination networks do not overlap. If the destination networks overlap, you will have one active and one standby.

Reply:
Combining this question with the previous one (a selector of 0.0.0.0/0). I would anticipate that a common pattern would be a pod with 2 egress networks where one is the "default" and the other hosts specific functionality on a given subnet. In this scenario the user would likely set EIP-A with a specific traffic selector (eg 1.2.3.0/24) and EIB-B with the default 0.0.0.0/0 to catch everything else. 

* Would these be treated as overlapping and fall into the one-active-one-standby scenario? 
* If so, would it be possible to allow at least this narrower use of overlap where one of the EgressIPTraffic contains the default route? Or could one EgressIPTraffic have an `isDefault: <true|false>` specifier?

Reply:
I see. Thanks for this insight from the users perspective.

This would complicate the implementation. We've have to have multiple logical router policy priorities to implement this.

> Would these be treated as overlapping and fall into the one-active-one-standby scenario?

Good question and this makes me rethink the existing one-active one standby and instead if there are overlapping networks (like your example), its simply undefined which one gets selected.
Id love to do what you suggested but it increases the complexity of the implementation that I am slow to agree to.

> If so, would it be possible to allow at least this narrower use of overlap where one of the EgressIPTraffic contains the default route? Or could one EgressIPTraffic have an isDefault: <true|false> specifier?

We could do this in a future RFE if its a limitation. If we added this, we'd need to add new logical route policies in OVN and new rule priorities in iproute2 ip rules.

Reply:
Follow up with customer did not raise any concerns.